### DISCUSSION TITLE
Copilot auto suggestion accuracy has taken a nose dive

### DISCUSSION THREAD

**Discussion by tmlewin** [community/community#19655] [Copilot Conversations]
I decided to test copilot against some of the leetcode questions it used to solve with ease a few months ago and the it keeps suggesting wrong answers with not even a sigle correct suggestion out of 8 questions. I noticed a massive degradation in the quality of output. I hope things really improve or atleast go back to how it was now it is paid service

**Comment by davecheney** [comment]
@tmlewin as an experiment go to https://github.com/settings/copilot and allow public code suggestions.

**Reply by tmlewin** [reply]
@davecheney  Hi it is already set to allow public code and still nothing improved. It took a big downturn in the accuracy of suggestions that i just turned it off. Even leetcode questions it used to solve with ease  , it can no longer provide a single correct suggestion to the same set of questions.

**Reply by hsenag** [reply]
Could you give an example of a question that's got worse?

**Reply by tmlewin** [reply]
It is total of 8 questions and counting .let me drop one of the challenges   

There are several cards arranged in a row, and each card has an associated number of points. The points are given in the integer array cardPoints.

In one step, you can take one card from the beginning or from the end of the row. You have to take exactly k cards.

Your score is the sum of the points of the cards you have taken.

Given the integer array cardPoints and the integer k, return the maximum score you can obtain.

 

Example 1:

Input: cardPoints = [1,2,3,4,5,6,1], k = 3
Output: 12
Explanation: After the first step, your score will always be 1. However, choosing the rightmost card first will maximize your total score. The optimal strategy is to take the three cards on the right, giving a final score of 1 + 6 + 5 = 12.
Example 2:

Input: cardPoints = [2,2,2], k = 2
Output: 4
Explanation: Regardless of which two cards you take, your score will always be 4.
Example 3:

Input: cardPoints = [9,7,7,9,7,7,9], k = 7
Output: 55
Explanation: You have to take all the cards. Your score is the sum of points of all cards.
 

Constraints:

1 <= cardPoints.length <= 105
1 <= cardPoints[i] <= 104
1 <= k <= cardPoints.length

**Reply by tmlewin** [reply]
@hsenag  second question. It used to solve these questions so easily in the past . But now it can't even make a single correct suggestion. I had to hand code the solution myself. I use leetcode to test the accuracy of suggestions. even in react projects it doesn't bring suggestions at all or sometimes when it does,  a high number is very irrelevant to what i am working on or inaccurate


Given an array nums with n integers, your task is to check if it could become non-decreasing by modifying at most one element.

We define an array is non-decreasing if nums[i] <= nums[i + 1] holds for every i (0-based) such that (0 <= i <= n - 2).

 

Example 1:

Input: nums = [4,2,3]
Output: true
Explanation: You could modify the first 4 to 1 to get a non-decreasing array.
Example 2:

Input: nums = [4,2,1]
Output: false
Explanation: You can't get a non-decreasing array by modify at most one element.
 

Constraints:

n == nums.length
1 <= n <= 104
-105 <= nums[i] <= 105

**Comment by davecheney** [comment]
@tmlewin looking at your settings, you appear to have _blocked_ public code results. I apologise for the language on the settings page, it is not as clear as it should be. Please try visiting https://github.com/settings/copilot, adjusting your settings if necessary, and if so, pay attention to the advice at the bottom of the page about restarting editors for any changes to take effect immediately. Thank you.

**Reply by tmlewin** [reply]
@davecheny. can you please check it now. It i not bblocked . I tried both allowed and blocked and it amounted to the same thing. I also tried both options after every restart of vscode and waited for the usual 30 minutes. I tried everything that was suggested at that link and the inaccuracies as i explained earlier is still there. It just took a downgrade from the accuracy of months ago.Thanks

**Reply by davecheney** [reply]
@tmlewin i can see that two minutes ago you flipped the setting from _block_ public code suggestions to _allow_ public code suggestions.

**Reply by tmlewin** [reply]
@davecheney  Yeah I was on  allow a few days ago it was not delivering. I blocked it samething. Then flipped it back to allow a few minutes ago. I have tried both options across the past days. Can u check my history. If u check my history , u can see that I have been flipping on both settings  for days and none has delivered . Same issue as explained above.

**Reply by davecheney** [reply]
I can see the setting has changed several times. We expect if public code suggestions are blocked the quality of suggestions will decrease. Beyond that, I don't know what to add I'm afraid.

**Reply by tmlewin** [reply]
@davecheney  i appreciate your effort. I just want to notify you guys that even with public code allowed , the quality of suggestions took a noticeable nosedive. 8 leetcode questions that was a breeze a few months ago . Now it cant even provide a single suggestion that works. same questions negative results in terms of accuracy.

**Comment by MulleDK19** [comment]
It's definitely become a lot worse. It suggests a lot of non-sense.
IE. I asked it to make a IsInPirateCostume function, and it made a function that checked if it was in a Parrot costume.
It used to be really useful for P/Invoke, but now it's a lot of non-sense, writing non-sense structs, with errors like COlor instead of Color, etc. (For something that has nothing to do with colors).


I did some testing. If I open up the Copilot window, it (sometimes) gets the original good code. The one it inserts when you just accept with Tab is far down the list of the generated solutions (if it's even there).

It seems that Copilot is now random and not favoring good solutions.

```csharp
using System;
using System.Runtime.InteropServices;

internal static class Program
{
    [DllImport("kernel32.dll", SetLastError = true)]
    private static extern bool SetCurrentConsoleFontEx(IntPtr consoleOutput, bool maximumWindow, ref CONSOLE_FONT_INFO_EX newFont);

    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
    private
```


Copilot suggests

```csharp
struct CONSOLE_FONT_INFO_EX
{
    public uint cbSize;
    public uint nFont;
    public COlorFontAttributeOptions dwFontSize;
    public int FontFamily;
    public int FontWeight;
    public ushort FaceName;
}
```

dwFontSize should be `COORDS`. It seems it's started generating that, then instead of `ORDS` generated `lorFontAttributeOptions` instead.

It's also made FaceName `ushort` which is inconvenient when it could be a string with a marshalling attribute.

Opening up Copilot instead of using Tab, gives the old, correct suggestion:

```csharp
struct CONSOLE_FONT_INFO_EX
{
    public uint cbSize;
    public uint nFont;
    public COORD dwFontSize;
    public int FontFamily;
    public int FontWeight;
    [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 32)]
    public string FaceName;
}
```

Sometimes, anyway.. Sometimes it only generates the bad solutions, and you have to re-generate until the right one appears. I even had to play with the count setting to get it to show sometimes.

Add a new line before the struct and its attribute: No good solution at all, ever..


With Copilot now suggesting code at random instead of the most fitting, it's no wonder it's taken a nose dive.



Some more examples, when the existing code is moved and reformatted a few times.

```csharp
struct CONSOLE_FONT_INFO_EX
{
    public uint cbSize;
    public uint nFont;
    public COORD dwFontSize;
    public int FontFamily;
    public int FontWeight;
    public short wcss;
    public short wc2;
    public short wc3;
    public short wc4;
    public short wc5;
    public short wc6;
}
```

```csharp
struct CONSOLE_FONT_INFO_EX
{
    public uint cbSize;
    public uint nFont;
    public COlorFontFaceW FontFace;
    public uint FontFamily;
    public uint FontWeight;
    public ushort FontSize;
    public ushort FontPitch;
    public uint FontCharSet;
    [MarshalAs(UnmanagedType.ByValArray, SizeConst = 32)]
    public ushort[] FaceName;
}
```

A month ago, I didn't even have to specify the StructLayout attribute, it would suggest that automatically. Now, without me specifying that attribute exactly like that, manually, it's even more non-sense:

```csharp
private struct CONSOLE_FONT_INFO_EX
{
    public uint cbSize;
    public uint nFont;
    public COORD dwFontSize;
    public int FontFamily;
    public int FontWeight;
    public ushort wBlackPixels;
    public ushort wWhitePixels;
    public uint nFontType;
    public uint nFontFamily;
    public uint nFontPitchAndFamily;
    public uint cCharacterSet;
    [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 32)]
    public string FaceName;
}
```

A month ago, it would still get less than ideal field types from time to time, but at least the types were valid and correct.

It doesn't help that the VS 2022 implementation is really shoddy, to a point where it's useless.

**Reply by tmlewin** [reply]
I am glad i am not the only one that noticed it. The copilot suggestion is bordeline useless to me at this point. It took a nose dive from an  amazing tool with incredible accuracy to a tool that can barely  give a suggestion with functional code that works. many stuff it could handle with ease before , it just cant solve it anymore.

I am on the trial period and if things dont improve i would simply  cancel the yearly subsription which i already planned on doing. I hope things improve

**Reply by MulleDK19** [reply]
The VS Code version also seems to use a different engine than the VS 2022 implementation. I get completely different suggestions depending on which one I use.

**Comment by MulleDK19** [comment]
@tmlewin Are you still using Copilot? Has it improved?
